browser-sign-in-kotlin/app/src/main/java/com/okta/browser/MainActivity.kt

```diff
@@ -0,0 +1,266 @@
+            override fun onCancel() {
+                network_progress.hide()
+                showMessage(getString(R.string.operation_cancelled))
+            }
+
+            override fun onError(msg: String?, exception: AuthorizationException?) {
+                signInError(msg, exception)
+            }
+
+            override fun onSuccess(result: AuthorizationStatus) {
+                network_progress.hide()
+                when (result) {
+                    AUTHORIZED -> signInSuccess()
+                    SIGNED_OUT -> showMessage(getString(R.string.sign_out_success))
+                }
+            }
+        }, this)
+    }
+
+    fun signOut() {
+        networkCallInProgress()
+        webAuthClient?.signOutOfOkta(this)
+    }
+
+    private fun signInSuccess() {
+        showMessage(getString(R.string.authorized))
+        supportFragmentManager.beginTransaction().replace(
+            R.id.fragment,
+            AuthorizedFragment.newInstance(customSignIn)
+        ).commit()
+    }
+
+    private fun signInError(msg: String?, exception: AuthorizationException?) {
+        network_progress.hide()
+        showMessage(msg ?: getString(R.string.unknown))
+        Log.d(tag, "onError: ", exception)
+    }
+
+    private fun signIn(
+        customSignIn: Boolean,
+        hintOrUsername: String,
+        password: String
+    ) {
+        if (customSignIn) {
+            hintOrUsername.isEmpty().or(password.isEmpty())
+                .run {
+                    if (this) {
+                        showMessage(getString(R.string.invalid_input))
+                    } else {
+                        networkCallInProgress()
+                        authenticateUser(hintOrUsername, password)
+                    }
+                }
+        } else {
+            networkCallInProgress()
+            webAuthClient?.signIn(this, AuthenticationPayload.Builder().setLoginHint(hintOrUsername).build())
+        }
+    }
+
+    private fun authenticateUser(username: String, password: String) {
+        GlobalScope.launch(Dispatchers.Main) {
+            withContext(Dispatchers.IO) {
+                authenticationClient?.authenticate(
+                    username, password.toCharArray(),
+                    null, null
+                )
+            }?.run {
+                authClient?.signIn(sessionToken, null, object : RequestCallback<Result, AuthorizationException> {
+                    override fun onSuccess(result: Result) {
+                        signInSuccess()
+                    }
+
+                    override fun onError(error: String?, exception: AuthorizationException?) {
+                        signInError(error, exception)
+                    }
+                })
+            }
+        }
+    }
+
+    fun getSession(): SessionClient? {
+        return if (customSignIn) {
+            authClient?.sessionClient
```

Comment:
The differences between `AuthClient` and `WebAuthClient` is the way it does login and log out. In the SDK it is different in that we need separate interfaces. But app devs can encapsulate the two in one `client`. But that is a trade off in memory since one client is not needed.

Reply:
Most app devs would just use one or the other entirely and we have both since we're a sample correct? I'm just thinking it would be cleaner for devs to read and use as a sample if we abstract the choice between the two as much as possible even if our sample app takes a slight memory hit to do it.

Reply:
Yes devs would use one or the other not both. Since this is only a sample, I'll change to init both clients and not worry about memory.